### PR TITLE
WCF breaking change

### LINKED ISSUE #34909: [Breaking change]: WCF Client RC 6.0 RC Release
### Description

WCF Client 6.0 RC Release.

Previously when using Duplex contract, failure to capture synchronization context on channel factory open.  The behavior now matches .NET framework, by default, we will now capture the synchronization context, and will dispatch callbacks to the captured context.  To revert to 4.x behavior, apply the `CallbackBehaviorAttribute` with the `UseSynchronizationContext` property set to `false`.

### Version

Other (please put exact version in description textbox)

### Previous behavior

Previously when using Duplex contract, failure to capture synchronization context on channel factory open.  

### New behavior

The behavior now matches .NET framework, by default, we will now capture the synchronization context, and will dispatch callbacks to the captured context. 

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [X] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

Bug fix [PR ](https://github.com/dotnet/wcf/pull/5044).  The new behavior now matches .NET framework.  

### Recommended action

 To revert to 4.x behavior, apply the CallbackBehaviorAttribute with the UseSynchronizationContext property set to false.

### Feature area

WCF Client

### Affected APIs

```
DuplexChannelFactory<TChannel>.Open
DuplexChannelFactory<TChannel>.BeginOpen/EndOpen
```

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #34909 

[Preview](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/wcf-client/6.0/duplex-synchronization-context?branch=pr-en-us-34914)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md](https://github.com/dotnet/docs/blob/c34cbe8336091068d68ca3b1ff3310d4545a1fdb/docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md) | [docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/wcf-client/6.0/duplex-synchronization-context?branch=pr-en-us-34914) |

### CODE REVIEW

File: docs/core/compatibility/toc.yml [hunk 1]
```diff
@@ -1432,6 +1432,8 @@ items:
       items:
       - name: .NET Standard 2.0 no longer supported
         href: wcf-client/6.0/net-standard-2-support.md
+      - name: Duplex contract captures synchronization context
```
Comment (mconnew) [inline docs/core/compatibility/toc.yml:1435]: DuplexChannelFactory captures synchronization context

File: docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+---
+title: "Breaking change: Duplex contract captures synchronization context"
```
Comment (mconnew) [inline docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md:2]:
```suggestion
title: "Breaking change: DuplexChannelFactory captures synchronization context"
```

File: docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+description: Learn about the breaking change in WCF Client 6.0 where the duplex contract now captures synchronization context when it's opened.
```
Comment (mconnew) [inline docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md:3]: description: Learn about the breaking change in WCF Client 6.0 where the DuplexChannelFactory now captures the synchronization context when it's opened.

File: docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ms.date: 04/06/2023
+---
+# Duplex contract captures synchronization context
+
+The duplex contract now captures the synchronization context when it's opened and dispatches callbacks to the captured context.
```
Comment (mconnew) [inline docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md:8]:
```suggestion
DuplexChannelFactory now captures the synchronization context when it's opened and dispatches callbacks to the captured context.
```

File: docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+## Previous behavior
+
+In previous versions, the duplex contract failed to capture synchronization context when the channel factory was opened.
```
Comment (mconnew) [inline docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md:12]: In previous versions, DuplexChannelFactory failed to capture the current synchronization context when the channel factory was opened.

File: docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+## New behavior
+
+Starting in WCF Client 6.0 RC, the duplex channel captures synchronization context and dispatches callbacks to the captured context. This behavior matches that of .NEt Framework.
```
Comment (mconnew) [inline docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md:16]:
```suggestion
Starting in WCF Client 6.0 RC, DuplexChannelFactory captures the current synchronization context when opened and dispatches callbacks to the captured context. This behavior matches that of .NET Framework.
```

File: docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+WCF Client 6.0 RC
+
+## Type of breaking change
+
+This change is a [behavioral change](../../categories.md#behavioral-change).
+
+## Reason for change
+
+This change was made to [fix a bug](https://github.com/dotnet/wcf/issues/4946) where the client hung when making an asynchronous call followed by a synchronous call. The new behavior matches .NET Framework functionality.
```
Comment (mconnew) [inline docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md:28]: That wasn't the bug for this change. It was an internal find which was reported directly to me. An issue was never opened for it. The internal problem was a callback was instantiating a STA COM object and as the initial callback and subsequent callbacks weren't being dispatched to a custom SynchronizationContext, the callbacks were happening on the wrong thread.

File: docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+## Recommended action
+
+If the new behavior is undesirable, you can revert to version 4.x behavior by applying the <xref:System.ServiceModel.CallbackBehaviorAttribute> to your contract with the <xref:System.ServiceModel.CallbackBehaviorAttribute.UseSynchronizationContext> property set to `false`.
```
Comment (mconnew) [inline docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md:32]:
```suggestion
If the new behavior is undesirable, you can revert to version 4.x behavior by applying the <xref:System.ServiceModel.CallbackBehaviorAttribute> to your callback contract with the <xref:System.ServiceModel.CallbackBehaviorAttribute.UseSynchronizationContext> property set to `false`.
```

**Review by mconnew — CHANGES_REQUESTED** [review]
Duplex contract should have been DuplexChannelFactory. I don't know if you want to put the generic parameter TChannel as part of the name or if it should be linked.

File: docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md (hunk 7 above)
Comment (gewarren) [inline docs/core/compatibility/wcf-client/6.0/duplex-synchronization-context.md:28]:
```suggestion
This change was made to fix a bug where callbacks were happening on the wrong thread. The new behavior matches .NET Framework functionality.
```

**Review by mconnew — APPROVED** [review]

**Pull request merged**